connection.go

```diff
@@ -1163,3 +1295,156 @@ func (conn *Connection) NewStream() (*Stream, error) {
+}
+
+// Unregister unregisters the connection watcher.
+func (w *connWatcher) Unregister() {
+	w.unregister.Do(func() {
+		close(w.done)
+	})
+	<-w.finished
+}
+
+// NewWatcher creates a new Watcher object for the connection.
+//
+// After watcher creation, the watcher callback is invoked for the first time.
+// In this case, the callback is triggered whether or not the key has already
+// been broadcast. All subsequent invocations are triggered with
+// box.broadcast() called on the remote host. If a watcher is subscribed for a
+// key that has not been broadcast yet, the callback is triggered only once,
+// after the registration of the watcher.
+//
+// The watcher callbacks are always invoked in a separate goroutine. A watcher
+// callback is never executed in parallel with itself, but they can be executed
+// in parallel to other watchers.
+//
+// If the key is updated while the watcher callback is running, the callback
+// will be invoked again with the latest value as soon as it returns.
+//
+// Watchers survive reconnection. All registered watchers are automatically
+// resubscribed when the connection is reestablished.
+//
+// Keep in mind that garbage collection of a watcher handle doesn’t lead to the
+// watcher’s destruction. In this case, the watcher remains registered. You
+// need to call Unregister() directly.
+//
+// Unregister() guarantees that there will be no the watcher's callback calls
+// after it, but Unregister() call from the callback leads to a deadlock.
+//
+// See:
+// https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_events/#box-watchers
+//
+// Since 1.10.0
+func (conn *Connection) NewWatcher(key string, callback WatchCallback) (Watcher, error) {
+	// TODO: check required features after:
+	//
+	// https://github.com/tarantool/go-tarantool/issues/120
+	var st chan watchState
+	// Get or create a shared data for the key.
+	if val, ok := conn.watchMap.Load(key); !ok {
+		st = make(chan watchState, 1)
+		st <- watchState{
+			value:   nil,
+			init:    true,
+			ack:     false,
+			cnt:     0,
+			changed: nil,
+		}
+
+		if val, ok := conn.watchMap.LoadOrStore(key, st); ok {
+			close(st)
+			st = val.(chan watchState)
+		}
+	} else {
+		st = val.(chan watchState)
+	}
+
+	state := <-st
+	// Send an initial watch request if needed.
+	if state.cnt == 0 {
+		if _, err := conn.Do(newWatchRequest(key)).Get(); err != nil {
+			st <- state
+			return nil, err
+		}
+		state.init = true
+		state.ack = true
+	}
+	state.cnt += 1
+	st <- state
+
+	// Start the watcher goroutine.
+	done := make(chan struct{})
+	finished := make(chan struct{})
+
+	go func() {
+		for {
+			state := <-st
+			if state.changed == nil {
+				state.changed = make(chan struct{})
+			}
+			st <- state
+
+			if !state.init {
+				callback(WatchEvent{
+					Conn:  conn,
+					Key:   key,
+					Value: state.value,
+				})
+
+				// Acknowledge the notification.
+				state = <-st
+				ack := state.ack
+				state.ack = true
```

Comment:
What is the purpose of `ack`?

Reply:
The idea is that a first goroutine that already done with a callback sends acknowledge to a Tarantool. The goroutine will switch the state.ack value from `false` to `true`:

```Go
				state = <-st
				ack := state.ack // save a previous value
				state.ack = true // it will be equivalent to `if !ack { state.ack = true }`
				st <- state

				if !ack {
					conn.Do(newWatchRequest(key)).Get()
				}
```

Reply:
> a first goroutine that already done with a callback sends acknowledge to a Tarantool

Does net.box also chooses the first one? It seems that picking the last one is a tricky task.

Reply:
As far as I understand, for now it's not used anywhere in Tarantool (and here)?

Reply:
> Does net.box also chooses the first one?

I think so.

> As far as I understand, for now it's not used anywhere in Tarantool (and here)?

What do you mean?

Reply:
> > As far as I understand, for now it's not used anywhere in Tarantool (and here)?
> 
> What do you mean?

Tarantool server does not care if there was or wasn't any acknowledge.

Reply:
> Tarantool server does not care if there was or wasn't any acknowledge.

It cares. It sends IPROTO_EVENT only after an acknowledge.

Reply:
Oh, thanks, I get it now.

This strategy should work well to process all event updates in chronological order in case of a single watcher, but is seems potentially dangerous in case of multiple watchers subscribed to single event. Or it should be alright?

Reply:
It looks like you are right: there are maybe a race condition: in case of too long callbacks. I've updated the implementation to avoid it.

Reply:
To be honest, I didn't understand what have changed in our implementation.